src/main/scala/wdlTools/generators/code/package.scala

```diff
@@ -44,6 +44,94 @@ trait Sized {
   def firstLineLength: Int = length
 }
 
+trait ExpressionState {
+  def canAdvanceTo(state: ExpressionState): Boolean
+}
+
+trait InitialState extends ExpressionState {
+  override def canAdvanceTo(state: ExpressionState): Boolean = {
+    state match {
+      case _: InStringState | _: InOperationState => true
+      case _                                      => false
+    }
+  }
+}
+
+/**
+  * The initial expression state.
+  */
+object StartState extends InitialState
+
+/**
+  * The state of being within a string expression but not within a placeholder.
+  * @param quoting the type of quoting for the string
+  */
+case class InStringState(quoting: Quoting.Quoting) extends ExpressionState {
+  override def canAdvanceTo(state: ExpressionState): Boolean = {
+    state match {
+      case _: InStringState   => true
+      case InPlaceholderState => true
+      case _                  => false
+    }
+  }
+}
+
+/**
+  * The state of being within a placeholder.
+  */
+object InPlaceholderState extends InitialState
```

Comment:
It looks like there is no transition from InPlaceholderState --> InPlaceholderState, yet placeholders can be nested. It looks like WdlGenerator.scala:599 could result in attempting that transition. 

The test nested_expr.wdl doesn't exactly test this case because the quotes will put it through string context first. The WDL spec says "Placeholders may contain other placeholders to any level of nesting, and placeholders are evaluated recursively in a depth-first manner."

Reply:
Ok, so nested use should be written as in the nested_expr.wdl test file.